### PR TITLE
AVRO-2893: Isolate Errors to Simplify Import Graph

### PR DESCRIPTION
**Pull request by kojiromike**
### Jira

- [x] My PR addresses [AVRO-2893](https://issues.apache.org/jira/browse/AVRO-2893)
- [x] No dependencies

### Tests

- [x] My PR does not add any new features requiring tests.

### Commits

- [x] My commits all reference Jira issues in their subject lines.

### Documentation

- [x] N/A

### CODE REVIEW

**Comment by kojiromike** [comment]
@Fokko I don't think Travis' failure has to do with these changes. Any suggestions?

**Comment by kojiromike** [comment]
@RyanSkraba how did you restart the build for just the one step that failed?

**Comment by RyanSkraba** [comment]
Hello!  There's a Restart Job button on https://travis-ci.org/github/apache/avro/jobs/712170660 (job 2252.3).

I added you as a reviewer to https://github.com/apache/avro/pull/939 :D  Let's get this deactivated while it's sending us a bad signal!

File: lang/py/avro/schema.py [hunk 1]
```diff
@@ -1120,22 +1100,25 @@ def make_avsc_object(json_data, names=None, validate_enum_symbols=True):
...
 # TODO(hammer): make method for reading from a file?
 
 
-def parse(json_string, validate_enum_symbols=True):
+def parse(json_string, validate_enum_symbols=True, register_codec=None):
     """Constructs the Schema from the JSON text.
 
+    @arg json_string: The json string of the schema to parse
     @arg validate_enum_symbols: If False, will allow enum symbols that are not valid Avro names.
+    @arg register_codec: If a string, registers a Python codec by that name.
```
Comment (RyanSkraba) [inline lang/py/avro/schema.py:1113]: Is including register_codec intentional for this PR?

File: lang/py/avro/errors.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+class ConnectionClosedException(AvroException):
+    """Raised when attempting IPC on a closed connection."""
+
+
+class ProtocolParseException(AvroException):
+    """Raised when a protocol failed to parse."""
+
+
+class AvroError(RuntimeError, AvroException):
```
Comment (RyanSkraba) [inline lang/py/avro/errors.py:78]: The doc for this is correct, it's only used for communicating with the user when a CLI tool is incorrectly parameterized.  How about a more specific name `AvroToolError` or `AvroParameterError` or `AvroUsageError`?

File: lang/py/avro/schema.py (hunk 1 above)
Comment (kojiromike) [reply lang/py/avro/schema.py:1113]: hah no, good catch

File: lang/py/avro/errors.py (hunk 2 above)
Comment (kojiromike) [reply lang/py/avro/errors.py]: Renamed to `UsageError`, since we're already in the namespace of avro.

**Comment by kojiromike** [comment]
@Fokko @RyanSkraba When I look at my own repos in travis-ci, I see buttons for restarting parts of the build, like this:

Screen Shot 2020-08-15 at 12 53 32 PM

…but I do not see those buttons when looking at apache/avro. Should I? It would be very nice if I could restart just parts of these builds.

**Pull request merged**